cmake/selective_build.cmake

```diff
@@ -0,0 +1,56 @@
+# Copyright (C) 2018-2020 Intel Corporation
+# SPDX-License-Identifier: Apache-2.0
+#
+function(ie_gen_default_header gen_header)
+    file(WRITE "${gen_header}" "#define GraphGen(...) 1")
+endfunction()
+
+function(ie_gen_header gen_header itt_traces itt_domains)
+    if(itt_traces)
+        file(WRITE "${gen_header}" "## This code is auto generated\n")
+        file(APPEND "${gen_header}" "#ifdef GraphGen\n")
+        file(APPEND "${gen_header}" "#error \"GraphGen already defined\"\n")
+        file(APPEND "${gen_header}" "#else\n")
+        file(APPEND "${gen_header}" "#define GraphGen(impl) impl\n")
+        file(APPEND "${gen_header}" "#endif\n")
+    file(GLOB files "${itt_traces}/*.csv")
+    LIST(LENGTH files FILESCOUNT)
+    if (FILESCOUNT LESS_EQUAL 0)
+        message(FATAL_ERROR "No traces are found in ${itt_traces}")
+    endif()
+    foreach(file ${files})
+        FILE(READ "${file}" contents)
+        STRING(REGEX REPLACE ";" "\\\\;" contents "${contents}")
+        STRING(REGEX REPLACE "\n" ";" contents "${contents}")
+        foreach(line ${contents})
+            STRING(REGEX REPLACE "," ";" line "${line}")
+            LIST(GET line 0 DOMAIN)
+            LIST(FIND itt_domains "${DOMAIN}" DOMAIN_FOUND)
+            if(DOMAIN_FOUND GREATER_EQUAL 0)
+                LIST(GET line 1 DEFINED_VALUE)
+                STRING(REGEX REPLACE "::" "_" DEFINED_VALUE "${DEFINED_VALUE}")
+                STRING(REGEX REPLACE "\\." "_" DEFINED_VALUE "${DEFINED_VALUE}")
+                set(DEFINED_VALUE "Gen_${DEFINED_VALUE}")
+                file(APPEND "${gen_header}" "#ifndef ${DEFINED_VALUE}\n")
+                file(APPEND "${gen_header}" "#define ${DEFINED_VALUE} 1\n")
+                file(APPEND "${gen_header}" "#endif\n")
+            endif()
+        endforeach()
+    endforeach()
+    else()
+        file(WRITE "${gen_header}" "#define GraphGen(...) 0")
+    endif()
+endfunction()
```

Comment:
BTW, how much time this function takes to execute? Looks like a lot of regex on each file line..